include/deal.II/grid/grid_tools.h

```diff
@@ -2142,6 +2162,453 @@ namespace GridTools
+                    // gradient (things tend to be locally flat, so the gradient
+                    // itself is usually small)
+                    if (std::abs(gradient_weight) > 10)
+                      {
+                        gradient_weight = -10.0;
+                        break;
+                      }
+                  }
+
+                // It only makes sense to take convex combinations with weights
+                // between zero and one. If the update takes us outside of this
+                // region then rescale the update to stay within the region and
+                // try again
+                Tensor<1, GeometryInfo<structdim>::vertices_per_cell> tentative_weights =
+                  guess_weights + gradient_weight*current_gradient;
+
+                double new_gradient_weight = gradient_weight;
+                for (unsigned int iteration_count = 0; iteration_count < 40; ++iteration_count)
+                  {
+                    if (weights_are_ok(tentative_weights))
+                      break;
+
+                    for (unsigned int i = 0; i < GeometryInfo<structdim>::vertices_per_cell; ++i)
+                      {
+                        if (tentative_weights[i] < 0.0)
+                          {
+                            tentative_weights -= (tentative_weights[i]/current_gradient[i])
+                                                 *current_gradient;
+                          }
+                        if (tentative_weights[i] < 0.0 || 1.0 < tentative_weights[i])
+                          {
+                            new_gradient_weight /= 2.0;
+                            tentative_weights = guess_weights + new_gradient_weight*current_gradient;
+                          }
+                      }
+                  }
+
+                // the update might still send us outside the valid region, so
+                // check again and quit if the update is still not valid
+                if (!weights_are_ok(tentative_weights))
+                  break;
+
+                // if we cannot get closer by traveling in the gradient direction then quit
+                if (get_point_from_weights(tentative_weights).distance(trial_point) <
+                    get_point_from_weights(guess_weights).distance(trial_point))
+                  guess_weights = tentative_weights;
+                else
+                  break;
+                Assert(weights_are_ok(guess_weights), ExcInternalError());
+              }
+            Assert(weights_are_ok(guess_weights), ExcInternalError());
+            projected_point =  get_point_from_weights(guess_weights);
+          }
+      }
+    else
+      {
+        Assert(false, ExcNotImplemented());
+        return projected_point;
+      }
+
+    // if structdim == 2 and the optimal point is not on the interior then
+    // we may be able to get a more accurate result by projecting onto the
+    // lines.
+    if (structdim == 2)
```

Comment:
you can only get here from line 2580. Maybe move the block from here to that place?

Reply:
fixed.